### PR TITLE
Introducing navigation across portal and microservices

### PR DESCRIPTION
**Pull request by einari**
This introduces the concept of having navigation that is consistent from the top level window to the inner iframe documents. This has been built with our convention of `/<microservice>` to `/_/<microservice>` for iframes in mind.

Changes that are introduced:
* Bootstrapper now includes a router - one does not need to set this up
* Switch is not used within Microservices for routing - as we need more control
* In the Portal we now have a new construct called `<CompositionRoute/>` that renders the correct iframe and deals with reloads and setting correct `src` for the iframe
* `<MicroserviceRoute/>` is used within a Microservice to tell what sub routes it is listening to. It does not need a full path as this is derived from the `MicroserviceContext``. It is optional whether or not this is used - depends on the microservice.
* A React Context is introduced; `MicroserviceContext` that holds the configuration given in the Bootstrapper and makes it available from anywhere - this is used to derive the prefix for any routes within a Microservice.
* `<Routing/>` is used within a Microservice and all `<MicroserviceRoute/>` instances lives within it. This governs the actual pushing of state to the iframe windows history.
* `NavigationItem` introduces a new optional property called `path`. Once this is set, navigation will work with the new `INavigator`. The path one puts into this is the top level path, not the one with the `/_` prefix - for anything within **Applications** this would be `/applications/something`.
* Message `NavigatedTo` gets published - putting in a dependency to `IMessenger` in a ViewModel lets you listen to this. Useful for those scenarios where the navigation is navigating data and not using the `<MicroserviceRoute/>`.

## How to test

1. Run environment according to [this](https://github.com/dolittle/Studio/blob/development/Documentation/getting-started.md)
2. Start portal Web as explained [here](https://github.com/dolittle/Studio/blob/development/Documentation/getting-started.md)
3. Start applications Web as explained [here](https://github.com/dolittle/Studio/blob/development/Documentation/getting-started.md)
4. Optionally start applications Backend as explained [here](https://github.com/dolittle/Studio/blob/development/Documentation/getting-started.md)
5. Navigate to [http://localhost:9000/applications](http://localhost:9000/applications) - you should see "This is the roooot" in the iframe content
6. Change the URL to [http://localhost:9000/applications/something](http://localhost:9000/applications/something) - you should see "Something" in the iframe content. (this is effectively same as refresh)

## Missing things

The biggest missing feature right now is probably that history navigation does not work (back and forward). We will have to revisit this.

The following tasks are known that is left overs / technical debt:
- Use navigator in the top level menu
- Navigation structure needs to set the correct item when loading based on current location path
- Fix back and forward for navigation
- Add a way to get a callback when navigation has occurred
- Add data and title to the navigator when navigating
- Add lifecycle events into the navigator (Before, After)

These are registered [here](https://app.asana.com/0/1190518762954630/1199621706420926)

## Challenge

The name `<MicroserviceRoute/>` feels wrong. We have a top level `<CompositionRoute/>` and I'd love a better name for whatever sits inside it. See this as a hierarchy.

### CODE REVIEW

**Review by jakhog — CHANGES_REQUESTED** [review]
Some minor suggestions, but one question:

How do we use this from within the iframe? I see how it's used with the `NavigationItem`s that is sent back to the portal, but is there a component to use for links within the frames? Or do we use the `INavigator` in `onclick` events for moving around?

**Comment by einari** [comment]
> Some minor suggestions, but one question:
> 
> How do we use this from within the iframe? I see how it's used with the `NavigationItem`s that is sent back to the portal, but is there a component to use for links within the frames? Or do we use the `INavigator` in `onclick` events for moving around?

That would've been helpful of me to include info on :) 
I didn't think too much about it being called directly from within a Microservice for the navigation - but that can be added as well. But in general the idea was that the Navigation system found in @shared/portal/navigation/Navigation - it has NavigationItems and on them you can now say which route/path to use for items that will be in the left most menu.

File: Source/Portal/Web/layouts/TopLevelMenuViewModel.ts [hunk 1]
```diff
@@ -2,12 +2,17 @@
...
 
 @injectable()
 export class TopLevelMenuViewModel {
-    constructor(private readonly _messenger: IMessenger) {
+    constructor(private readonly _navigator: INavigator, private readonly _messenger: IMessenger) {
+    }
+
+    doMagic() {
+        this._navigator.navigateTo('/applications/something');
     }
```
[the remark this answers is not in the document]
Comment (einari) [reply Source/Portal/Web/layouts/TopLevelMenuViewModel.ts:16]: Thanks.. Good catch - leftovers indeed.

File: Source/Shared/Web/routing/RoutingViewModel.tsx [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+@injectable()
+export class RoutingViewModel {
+    currentPath: string;
+
+    navigated!: (path: string) => void;
+
+    constructor(readonly microserviceConfiguration: MicroserviceConfiguration, private readonly _messenger: IMessenger) {
+        this.currentPath = microserviceConfiguration.prefix;
+        _messenger.subscribeTo(NavigatedTo, _ => {
+            const actualPath = `/_${_.path}`;
```
[the remark this answers is not in the document]
Comment (einari) [reply Source/Shared/Web/routing/RoutingViewModel.tsx:18]: I see a pattern emerging :)

File: Source/Shared/WebPack/frontend/devServer.js [hunk 3]
```diff
@@ -11,6 +13,17 @@ module.exports = (basePath) => {
...
+        before: (app, server, compiler) => {
+            app.get('*', (req, res, next) => {
+                const match = req.originalUrl.match(fileTypes);
+                if (match && match.length > 0) {
+                    next();
+                    return;
+                }
+                const html = require('./HtmlInterceptorPlugin').getGeneratedHtml();
+                res.send(html);
+            });
```
[the remark this answers is not in the document]
Comment (einari) [reply Source/Shared/WebPack/frontend/devServer.js:26]: WebPack DevServer does not seem to function properly with HistoryAPIFallback for DevServers that has a basePath. It didn't fallback to index.html for all routes as it should. So I implemented a fallback that serves the finished rendered index.html. To be able to do that I had to create a plugin that uses the WebPack hooks and gets called when the index.ejs has been "rendered" to index.html and keep it in memory. Also known as "trickster in Ludo"

[the remark this answers is not in the document]
Comment (jakhog) [reply Source/Shared/WebPack/frontend/devServer.js:26]: Aha, yes that's a real Ludo trick you got there 😛

File: Source/Shared/Web/routing/MicroserviceRoute.tsx [hunk 4]
```diff
@@ -14,10 +14,14 @@ export const MicroserviceRoute = (props: RouteProps) => {
...
-                innerProps.path = innerProps.path.replace('//', '/');
 
-                console.log(`Route path : ${innerProps.path}`);
+                let path = `${props.path} || ''`;
+                if (path.startsWith('/')) {
+                    path = path.substr(1);
+                }
+
+                innerProps.path = `${value.prefix}/${path}`;
+                innerProps.path = innerProps.path.replace('//', '/');
```
Comment (jakhog) [inline Source/Shared/Web/routing/MicroserviceRoute.tsx:24]: Now we shouldn't need this right?
```suggestion
```

File: Source/Shared/Web/routing/RoutingViewModel.tsx [hunk 5]
```diff
@@ -15,8 +15,12 @@ export class RoutingViewModel {
...
         this.currentPath = microserviceConfiguration.prefix;
         _messenger.subscribeTo(NavigatedTo, _ => {
-            const actualPath = `/_${_.path}`;
-            console.log(`Navigated to ${actualPath}`);
+            let path = _.path;
+            if (path.startsWith('/')) {
+                path = path.substr(1);
+            }
+
+            const actualPath = `/_${path}`;
```
Comment (jakhog) [inline Source/Shared/Web/routing/RoutingViewModel.tsx:23]: And we need a `/` here:
```suggestion
            const actualPath = `/_/${path}`;
```

**Review by jakhog — COMMENTED** [review]
That makes sense, and for now I don't think we need any special thing for navigating from inside the frames. So should we leave that as an exercise for later?

**Comment by einari** [comment]
Merged and we now have navigation from menu with deep linking :)

**Pull request merged**